sympy/simplify/cse_main.py

```diff
@@ -536,3 +536,101 @@ def cse(exprs, symbols=None, optimizations=None, postprocess=None,
         return replacements, reduced_exprs
 
     return postprocess(replacements, reduced_exprs)
+
+
+class Marker:
+    pass
+
+# XXX: Should this go in sympy.utilities.iterables?
+def shortest_repeated_subsequence(S, ignore=(Marker,)):
```

Comment:
It also seems to correctly handle repeated products, which is good

```
>>> cse_strs(['xxxx'])
Starting ops 3
Ending ops 1 + 1 = 2
(['AA'], {'A': 'xx'})
>>> cse_strs(['xxxxx'])
Starting ops 4
Ending ops 2 + 1 = 3
(['AAx'], {'A': 'xx'})
>>> cse_strs(['xxxxxx'])
Starting ops 5
Ending ops 1 + 2 = 3
(['BA'], {'A': 'xx', 'B': 'AA'})
```

Note that [x, x] is indeed the shortest repeated subsequence in [x, x, x, x] by the definition in the docstring because we can take the subsequence starting from the second to last element, [x, x, **x, x**] and it can't be extended.